### PR TITLE
Bug 1217710 - [TV Browser] Reorgnise the synced bookmark folder structure as Fennec did. r=seanlee

### PR DESCRIPTION
**Pull request by danhuang1202**
…nt between bookmarks from desktop and mobile, r=yifan

### CODE REVIEW

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=763dcc9103965555fd8cc437bd01140475a4d41c

File: tv_apps/browser/js/bookmarkStore.js [hunk 1]
```diff
@@ -107,12 +112,26 @@
               if(syncBookmark.length > 0) {
                 this.cache.push({
                   id: SYNC_BOOKMARK_ROOT_FOLDER_ID,
-                  // XXX we need UX's suggestion to have a folder name for root
-                  // folder and don't forget to handle l10n case.
-                  title: 'Synced Bookmarks',
+                  title: navigator.mozL10n.get('fxsync-desktop-bookmarks'),
```
Comment (begeeben) [inline tv_apps/browser/js/bookmarkStore.js]: Please use ```formatValue``` instead of ```get```.

File: tv_apps/browser/js/bookmarkStore.js [hunk 2]
```diff
@@ -107,12 +112,26 @@
...
+
+                // If there is mobile folder and also has children in it,
+                // create a folder item to show the synced mobile bookmark
+                var length = syncBookmark.length;
+                for(var i = 0; i < length; i++) {
+                  if(syncBookmark[i].id === SYNC_BOOKMARK_MOBILE_FOLDER_ID &&
+                    syncBookmark[i].children.length > 0) {
+                    this.cache.push({
+                      id: SYNC_BOOKMARK_MOBILE_FOLDER_ID,
+                      title: navigator.mozL10n.get('fxsync-mobile-bookmarks'),
```
Comment (begeeben) [inline tv_apps/browser/js/bookmarkStore.js]: Please use ```formatValue``` instead of ```get```.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=b0a012934bed9dc4fa8c7bedc162e05a01001090

File: tv_apps/browser/js/bookmarkStore.js [hunk 3]
```diff
@@ -99,33 +141,50 @@
...
               // folder item to show the synced bookmark.
-              if(syncBookmark.length > 0) {
-                this.cache.push({
-                  id: SYNC_BOOKMARK_ROOT_FOLDER_ID,
-                  // XXX we need UX's suggestion to have a folder name for root
-                  // folder and don't forget to handle l10n case.
-                  title: 'Synced Bookmarks',
-                  type: 'folder',
-                  readOnly: true
+              if(syncPlacesBookmark.length > 0) {
```
Comment (weilonge) [inline tv_apps/browser/js/bookmarkStore.js]: If 'mobile' folder is not in syncPlacesBookmark, it is still possible that there is a 'mobile' folder record in SyncBrowserDB. Please consider the case for mobile folder separately.

File: tv_apps/browser/js/bookmarkStore.js [hunk 4]
```diff
@@ -92,6 +129,11 @@
             resolve();
           });
         } else {
+//IFNDEF_FIREFOX_SYNC
+        resolve();
```
Comment (weilonge) [inline tv_apps/browser/js/bookmarkStore.js]: Incorrect indent. add two spaces for this line.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=be461d16e5f0668a770f3bae69243e32968bf126

File: tv_apps/browser/js/bookmarkStore.js [hunk 5]
```diff
@@ -19,6 +24,38 @@
...
+      l10nPromises
+        .push(navigator.mozL10n.formatValue('fxsync-desktop-bookmarks'));
+      l10nPromises
+        .push(navigator.mozL10n.formatValue('fxsync-mobile-bookmarks'));
+      Promise.all(l10nPromises).then(
+        ([l10nFxsyncDesktopBookmarks, l10nFxsyncMobileBookmarks]) => {
+          this.syncBookmarkRootFolderCache.title = l10nFxsyncDesktopBookmarks;
+          this.syncBookmarkMobileFolderCache.title = l10nFxsyncMobileBookmarks;
+        }
+      );
```
Comment (weilonge) [inline tv_apps/browser/js/bookmarkStore.js]: Please add return Promise.all here. let's make init function supports Promise.

File: tv_apps/browser/js/bookmarkStore.js [hunk 6]
```diff
@@ -19,6 +24,38 @@
...
+//ENDIF_FIREFOX_SYNC
```
Comment (weilonge) [inline tv_apps/browser/js/bookmarkStore.js]: Handle the case of disabling Firefox Sync.
```js
//IFNDEF_FIREFOX_SYNC
          return Promise.resolve();
//ENDIF_FIREFOX_SYNC
```

File: tv_apps/browser/js/bookmarkStore.js [hunk 7]
```diff
@@ -82,6 +119,70 @@
...
+          }, syncMobileBookmarks => {
+            resolve(syncMobileBookmarks);
+          });
+        }),
+        new Promise(resolve => {
+          // get bookmark data from origin indexdDB
+          BrowserDB.getBookmarks(localBookmarks => {
+            resolve(localBookmarks);
+          });
+        })
```
Comment (weilonge) [inline tv_apps/browser/js/bookmarkStore.js]: Based on the [code](https://github.com/mozilla-b2g/gaia/pull/32797/files#diff-ed4ae19263bf74556f7542b37c38dbefR204), it is necessary to keep the execution sequence.
Please use 'then' to handle these three cases.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=a47660f9a865eee9214ae9dd7ac48747b2cd741e

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=4725097afa5cc20aae980c9c4d1535261254e1b6

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=191206745f5a66b0e63fd6151a99285b60f85f2a

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=8443f983a94b83af2077a27a4852875546bb88b6

**Pull request merged**